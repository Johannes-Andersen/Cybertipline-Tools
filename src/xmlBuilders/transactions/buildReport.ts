import { XMLBuilder, XMLParser } from 'fast-xml-parser';
import type { Report } from '../../types';
import { buildCellPhoneIncident } from '../incidents/buildCellPhoneIncident';
import { buildChatImIncident } from '../incidents/buildChatImIncident';
import { buildEmailIncident } from '../incidents/buildEmailIncident';
import { buildNewsgroupIncident } from '../incidents/buildNewsgroupIncident';
import { buildOnlineGamingIncident } from '../incidents/buildOnlineGamingIncident';
import { buildWebPageIncident } from '../incidents/buildWebPageIncident';
import { buildIncidentSummary } from '../reportDetails/buildIncidentSummary';
import { buildIntendedRecipient } from '../reportDetails/buildIntendedRecipient';
import { buildLawEnforcement } from '../reportDetails/buildLawEnforcement';
import { buildPersonOrUserReported } from '../reportDetails/buildPersonOrUserReported';
import { buildReporter } from '../reportDetails/buildReporter';
import { buildVictim } from '../reportDetails/buildVictim';

const parser = new XMLParser({
  attributeNamePrefix: '@_',
  ignoreAttributes: false,
  allowBooleanAttributes: true,
});

const builder = new XMLBuilder({
  attributeNamePrefix: '@_',
  ignoreAttributes: false,
  suppressBooleanAttributes: false,
  suppressEmptyNode: true,
  processEntities: false,
});

export const buildReport = (report: Report, keyName = 'report'): string => {
  const {
    victim,
    intendedRecipient,
    personOrUserReported,
    reporter,
    lawEnforcement,
    incidentDetails,
    incidentSummary,
    additionalInfo,
  } = report;

  const victims = victim?.map((e) => parser.parse(buildVictim(e)).victim);
  const intendedRecipients = intendedRecipient?.map(
    (e) => parser.parse(buildIntendedRecipient(e)).intendedRecipient,
  );
  const reportedPerson = personOrUserReported
    ? parser.parse(buildPersonOrUserReported(personOrUserReported))
        .personOrUserReported
    : undefined;
  const reportingPerson = parser.parse(buildReporter(reporter)).reporter;
  const lawEnforcementEntity = lawEnforcement
    ? parser.parse(buildLawEnforcement(lawEnforcement)).lawEnforcement
    : undefined;
  const incidentSummaryEntity = parser.parse(
    buildIncidentSummary(incidentSummary),
  ).incidentSummary;

  const {
    webPageIncidents,
    emailIncidents,
    newsgroupIncidents,
    chatImIncidents,
    onlineGamingIncidents,
    cellPhoneIncidents,
  } = incidentDetails ?? {};

  const internetDetailsEntities = [
    webPageIncidents?.map((e) => parser.parse(buildWebPageIncident(e))),
    emailIncidents?.map((e) => parser.parse(buildEmailIncident(e))),
    newsgroupIncidents?.map((e) => parser.parse(buildNewsgroupIncident(e))),
    chatImIncidents?.map((e) => parser.parse(buildChatImIncident(e))),
    onlineGamingIncidents?.map((e) =>
      parser.parse(buildOnlineGamingIncident(e)),
    ),
    cellPhoneIncidents?.map((e) => parser.parse(buildCellPhoneIncident(e))),
  ];

  const xmlObj = {
    '?xml': {
      '@_version': '1.0',
      '@_encoding': 'UTF-8',
    },
    [keyName]: {
      '@_xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance',
      '@_xsi:noNamespaceSchemaLocation':
        'https://report.cybertip.org/ispws/xsd',
      incidentSummary: incidentSummaryEntity,
      internetDetails: internetDetailsEntities.flat(),
      lawEnforcement: lawEnforcementEntity,
      reporter: reportingPerson,
      personOrUserReported: reportedPerson,
      intendedRecipient: intendedRecipients,
      victim: victims,
      additionalInfo,
    },
  };

  return builder.build(xmlObj);
};
