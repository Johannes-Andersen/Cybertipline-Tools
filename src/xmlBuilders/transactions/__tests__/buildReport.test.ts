import { describe, expect, it } from 'vitest';
import {
  Country,
  DeviceEventName,
  DeviceIdType,
  IPEventName,
  IncidentType,
  ReportAnnotation,
  State,
} from '../../../types';
import type { Report } from '../../../types';
import { buildReport } from '../buildReport';

describe('buildReport', () => {
  it('should build a report with all fields and multiple incident types', () => {
    const report: Report = {
      incidentSummary: {
        incidentType: IncidentType.ChildPornography,
        escalateToHighPriority: 'Immediate risk to child',
        reportAnnotations: {
          [ReportAnnotation.Sextortion]: true,
          [ReportAnnotation.CsamSolicitation]: true,
        },
        incidentDateTime: new Date('2024-01-15T14:30:00Z'),
        incidentDateTimeDescription: 'Multiple incidents over past week',
      },
      incidentDetails: {
        newsgroupIncidents: [
          {
            additionalInfo: 'Multiple incidents reported over the past week',
            content: 'Harmful content shared via newsgroup',
            name: 'newsgroup1',
          },
        ],
        webPageIncidents: [
          {
            url: ['https://example.com/harmful-content'],
            thirdPartyHostedContent: true,
            additionalInfo: 'Content reported by multiple users',
          },
        ],
        chatImIncidents: [
          {
            chatClient: 'SecureChat',
            chatRoomName: 'private-room-123',
            content: 'Suspicious messages exchanged',
            additionalInfo:
              'Room was created specifically for this interaction',
          },
        ],
        emailIncidents: [
          {
            emailAddress: [
              {
                email: 'suspicious@example.com',
                verified: true,
              },
            ],
            content: 'Harmful content shared via email',
            additionalInfo: 'Multiple emails sent from this address',
          },
        ],
      },
      lawEnforcement: {
        agencyName: 'FBI Cyber Division',
        caseNumber: 'FBI-2024-001',
        officerContact: {
          firstName: 'John',
          lastName: 'Smith',
          email: [
            {
              email: 'john.smith@fbi.gov',
              verified: true,
            },
          ],
        },
        reportedToLe: true,
        servedLegalProcessDomestic: true,
      },
      reporter: {
        reportingPerson: {
          firstName: 'Sarah',
          lastName: 'Johnson',
          email: [
            {
              email: 'sarah.johnson@company.com',
              verified: true,
            },
          ],
          phone: [
            {
              number: '555-0123',
              verified: true,
            },
          ],
        },
        companyTemplate: 'Standard reporting template for Company XYZ',
        termsOfService: 'Users must comply with all applicable laws',
      },
      personOrUserReported: {
        personOrUserReportedPerson: {
          firstName: 'Unknown',
          lastName: 'User',
        },
        screenName: 'suspicious_user_123',
        ipCaptureEvent: [
          {
            ipAddress: '192.168.1.100',
            eventName: IPEventName.Upload,
            dateTime: new Date('2024-01-15T12:00:00Z'),
          },
        ],
        deviceId: [
          {
            idType: DeviceIdType.IMEI,
            idValue: '123456789012345',
            eventName: DeviceEventName.Login,
          },
        ],
      },
      intendedRecipient: [
        {
          screenName: 'victim_user',
          espIdentifier: 'USER123',
          estimatedLocation: {
            city: 'Springfield',
            region: 'Illinois',
            countryCode: Country.UnitedStates,
          },
        },
      ],
      victim: [
        {
          victimPerson: {
            firstName: 'Jane',
            lastName: 'Doe',
            age: 15,
            address: [
              {
                address: '123 Main St',
                city: 'Springfield',
                state: State.Illinois,
                zipCode: '62701',
              },
            ],
          },
          schoolName: 'Springfield High School',
        },
      ],
      additionalInfo: 'Multiple incidents reported over the past week',
    };

    const result = buildReport(report);

    expect(result).toMatchSnapshot();
  });

  it('should build a report with only required fields', () => {
    const report: Report = {
      incidentSummary: {
        incidentType: IncidentType.OnlineEnticement,
        incidentDateTime: new Date('2024-01-16T09:15:00Z'),
      },
      reporter: {
        reportingPerson: {
          firstName: 'John',
          lastName: 'Smith',
          email: [
            {
              email: 'john.smith@company.com',
              verified: true,
            },
          ],
        },
      },
    };

    const result = buildReport(report);

    expect(result).toMatchSnapshot();
  });

  it('should build a report with multiple victims and recipients', () => {
    const report: Report = {
      incidentSummary: {
        incidentType: IncidentType.ChildSexTrafficking,
        incidentDateTime: new Date('2024-01-17T10:00:00Z'),
        escalateToHighPriority: 'Multiple victims identified',
      },
      reporter: {
        reportingPerson: {
          firstName: 'Emily',
          lastName: 'Davis',
          email: [
            {
              email: 'emily.davis@org.com',
              verified: true,
            },
          ],
        },
      },
      victim: [
        {
          victimPerson: {
            firstName: 'Victim',
            lastName: 'One',
            age: 14,
          },
          schoolName: 'Middle School A',
        },
        {
          victimPerson: {
            firstName: 'Victim',
            lastName: 'Two',
            age: 15,
          },
          schoolName: 'Middle School B',
        },
      ],
      intendedRecipient: [
        {
          screenName: 'user1',
          espIdentifier: 'USER456',
        },
        {
          screenName: 'user2',
          espIdentifier: 'USER789',
        },
      ],
    };

    const result = buildReport(report);

    expect(result).toMatchSnapshot();
  });

  it('should build a report with mixed incident types and locations', () => {
    const report: Report = {
      incidentSummary: {
        incidentType: IncidentType.ChildPornography,
        incidentDateTime: new Date('2024-01-18T15:45:00Z'),
      },
      reporter: {
        reportingPerson: {
          firstName: 'Michael',
          lastName: 'Wilson',
          email: [
            {
              email: 'michael.wilson@company.com',
              verified: true,
            },
          ],
        },
      },
      incidentDetails: {
        webPageIncidents: [
          {
            url: [
              'https://platform1.com/content',
              'https://platform2.com/content',
            ],
            additionalInfo: 'Content found on multiple platforms',
          },
        ],
        onlineGamingIncidents: [
          {
            gameName: 'Virtual World',
            console: 'PC Gaming',
            content: 'Inappropriate messages in game chat',
          },
        ],
        cellPhoneIncidents: [
          {
            phoneNumber: {
              number: '555-0123',
              verified: true,
            },
            latitude: 40.7128,
            longitude: -74.006,
          },
        ],
      },
      personOrUserReported: {
        screenName: 'multi_platform_user',
        deviceId: [
          {
            idType: DeviceIdType.IMEI,
            idValue: '987654321098765',
            eventName: DeviceEventName.Login,
          },
        ],
        estimatedLocation: {
          city: 'New York',
          region: 'New York',
          countryCode: Country.UnitedStates,
          verified: true,
        },
      },
    };

    const result = buildReport(report);

    expect(result).toMatchSnapshot();
  });
});
