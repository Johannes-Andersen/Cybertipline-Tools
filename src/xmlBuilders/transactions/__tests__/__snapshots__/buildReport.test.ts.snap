// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildReport > should build a report with all fields and multiple incident types 1`] = `"<?xml version="1.0" encoding="UTF-8"?><report xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://report.cybertip.org/ispws/xsd"><incidentSummary><incidentType>Child Pornography (possession, manufacture, and distribution)</incidentType><escalateToHighPriority>Immediate risk to child</escalateToHighPriority><reportAnnotations><sextortion/><csamSolicitation/></reportAnnotations><incidentDateTime>2024-01-15T14:30:00.000Z</incidentDateTime><incidentDateTimeDescription>Multiple incidents over past week</incidentDateTimeDescription></incidentSummary><internetDetails><webPageIncident thirdPartyHostedContent="true"><url>https://example.com/harmful-content</url><additionalInfo>Content reported by multiple users</additionalInfo></webPageIncident></internetDetails><internetDetails><emailIncident><emailAddress verified="true">suspicious@example.com</emailAddress><content>Harmful content shared via email</content><additionalInfo>Multiple emails sent from this address</additionalInfo></emailIncident></internetDetails><internetDetails><newsgroupIncident><name>newsgroup1</name><content>Harmful content shared via newsgroup</content><additionalInfo>Multiple incidents reported over the past week</additionalInfo></newsgroupIncident></internetDetails><internetDetails><chatImIncident><chatClient>SecureChat</chatClient><chatRoomName>private-room-123</chatRoomName><content>Suspicious messages exchanged</content><additionalInfo>Room was created specifically for this interaction</additionalInfo></chatImIncident></internetDetails><lawEnforcement><agencyName>FBI Cyber Division</agencyName><caseNumber>FBI-2024-001</caseNumber><officerContact><firstName>John</firstName><lastName>Smith</lastName><email verified="true">john.smith@fbi.gov</email></officerContact><reportedToLe>true</reportedToLe><servedLegalProcessDomestic>true</servedLegalProcessDomestic></lawEnforcement><reporter><reportingPerson><firstName>Sarah</firstName><lastName>Johnson</lastName><phone verified="true">555-0123</phone><email verified="true">sarah.johnson@company.com</email></reportingPerson><companyTemplate>Standard reporting template for Company XYZ</companyTemplate><termsOfService>Users must comply with all applicable laws</termsOfService></reporter><personOrUserReported><personOrUserReportedPerson><firstName>Unknown</firstName><lastName>User</lastName></personOrUserReportedPerson><screenName>suspicious_user_123</screenName><ipCaptureEvent><ipAddress>192.168.1.100</ipAddress><eventName>Upload</eventName><dateTime>2024-01-15T12:00:00.000Z</dateTime></ipCaptureEvent><deviceId><idType>IMEI</idType><idValue>123456789012345</idValue><eventName>Login</eventName></deviceId></personOrUserReported><intendedRecipient><espIdentifier>USER123</espIdentifier><screenName>victim_user</screenName><estimatedLocation><city>Springfield</city><region>Illinois</region><countryCode>US</countryCode></estimatedLocation></intendedRecipient><victim><victimPerson><firstName>Jane</firstName><lastName>Doe</lastName><address><address>123 Main St</address><city>Springfield</city><zipCode>62701</zipCode><state>IL</state></address><age>15</age></victimPerson><schoolName>Springfield High School</schoolName></victim><additionalInfo>Multiple incidents reported over the past week</additionalInfo></report>"`;

exports[`buildReport > should build a report with mixed incident types and locations 1`] = `"<?xml version="1.0" encoding="UTF-8"?><report xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://report.cybertip.org/ispws/xsd"><incidentSummary><incidentType>Child Pornography (possession, manufacture, and distribution)</incidentType><reportAnnotations/><incidentDateTime>2024-01-18T15:45:00.000Z</incidentDateTime></incidentSummary><internetDetails><webPageIncident><url>https://platform1.com/content</url><url>https://platform2.com/content</url><additionalInfo>Content found on multiple platforms</additionalInfo></webPageIncident></internetDetails><internetDetails><onlineGamingIncident><gameName>Virtual World</gameName><console>PC Gaming</console><content>Inappropriate messages in game chat</content></onlineGamingIncident></internetDetails><internetDetails><cellPhoneIncident><phoneNumber verified="true">555-0123</phoneNumber><latitude>40.7128</latitude><longitude>-74.006</longitude></cellPhoneIncident></internetDetails><reporter><reportingPerson><firstName>Michael</firstName><lastName>Wilson</lastName><email verified="true">michael.wilson@company.com</email></reportingPerson></reporter><personOrUserReported><screenName>multi_platform_user</screenName><deviceId><idType>IMEI</idType><idValue>987654321098765</idValue><eventName>Login</eventName></deviceId><estimatedLocation verified="true"><city>New York</city><region>New York</region><countryCode>US</countryCode></estimatedLocation></personOrUserReported></report>"`;

exports[`buildReport > should build a report with multiple victims and recipients 1`] = `"<?xml version="1.0" encoding="UTF-8"?><report xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://report.cybertip.org/ispws/xsd"><incidentSummary><incidentType>Child Sex Trafficking</incidentType><escalateToHighPriority>Multiple victims identified</escalateToHighPriority><reportAnnotations/><incidentDateTime>2024-01-17T10:00:00.000Z</incidentDateTime></incidentSummary><reporter><reportingPerson><firstName>Emily</firstName><lastName>Davis</lastName><email verified="true">emily.davis@org.com</email></reportingPerson></reporter><intendedRecipient><espIdentifier>USER456</espIdentifier><screenName>user1</screenName></intendedRecipient><intendedRecipient><espIdentifier>USER789</espIdentifier><screenName>user2</screenName></intendedRecipient><victim><victimPerson><firstName>Victim</firstName><lastName>One</lastName><age>14</age></victimPerson><schoolName>Middle School A</schoolName></victim><victim><victimPerson><firstName>Victim</firstName><lastName>Two</lastName><age>15</age></victimPerson><schoolName>Middle School B</schoolName></victim></report>"`;

exports[`buildReport > should build a report with only required fields 1`] = `"<?xml version="1.0" encoding="UTF-8"?><report xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://report.cybertip.org/ispws/xsd"><incidentSummary><incidentType>Online Enticement of Children for Sexual Act</incidentType><reportAnnotations/><incidentDateTime>2024-01-16T09:15:00.000Z</incidentDateTime></incidentSummary><reporter><reportingPerson><firstName>John</firstName><lastName>Smith</lastName><email verified="true">john.smith@company.com</email></reportingPerson></reporter></report>"`;
